### ISSUE TITLE
Prompt in the shell (pun intended)

### ISSUE THREAD

**Issue by JoseForest** [mugiwara85/CodeblockCustomizer#112]
Great plugins !

Could you add the possibility to add a prompt in the Code Block so it can be inserted at the beginning of each line of the block. I'm thinking especially for bash or other shell type code blocks when showing command line functions and their results. 

The prompt could be enable/disabled, be a default $ or defined with a string of characters. It could be skipped for some lines when it's not a prompt line.

Something that would look like this for the result (the prompt being defined as "user_name@host:\~$"):

```bash
user_name@host:~$ pwd
/home/user_name

user_name@host:~$
```

A more complex request: If the prompt include a path, and the command change that path, would it be possible to modify the prompt within a code block?

Again, something that would look like this for the result:

```bash
user_name@host:~$ cd test

user_name@host:~/test$
```
Here, the prompt is defined as "user_name@host:\~$" at the beginning of the code block but it's modified to "user_name@host:\~/test$" later in the block.

It would basically require a way to define the prompt string within the code block.

Thanks

**Comment by mugiwara85** [comment]
To answer you first question:
yes, that would be possible. I also considered this, I just haven't had time to implement it. The only problem here could be on how to define which line should or should not have the prompt displayed at the beginning. I could probably make a toggle, something like `Always show prompts`. 
 - When enabled then all lines have the prompt displayed, but you could disable it for specific line using a parameter like `noprompt` and defining the line numbers. 
 - When the option is disabled, then the prompt is not displayed, but you could show it for specific lines using the `prompt` parameter and defining the line numbers. 

You could define the prompt itself. For example let's say you'd like the prompt `user_name@host:~$`. Then you'd have to define this first as prompt1, and then define the languages, for which you want to use it. For example `bash`. You could then define a second prompt with `PS` for `Powershell`, or `>>>` for `python`.

Would this work for you? Or do you have a better idea?

To answer you second question:
Maybe. I don't want to promise anything for now, but I'll try to solve it.

**Comment by JoseForest** [comment]
Thanks for your fast reply.

I was thinking more like the way you define title, line numbers and highlighted lines or text, in the header of the code block.

For the specific code block where we want a prompt, we would include in the header something like 
` ```bash prompt:true ln:false` to get a default '$' prompt and something like : ` ```bash prompt:"user_name@host:~$" ln:false ` to get a specific prompt.

By default, the prompt would be disabled. I believe the need for prompt is not that high to justify a default enabled prompt for code blocks. But if you think otherwise, it could be a toggle to enable or disable it by default but still using 'true' ou 'false' in the header to override it for each block, like you do with the line numbers (if the prompt is defined with a string it means it's true for that block).

For the lines where we don't want to display a prompt, in its plugin, Mr Park, [https://github.com/HeekangPark/obsidian-hk-code-block](url) is using a special character at the end of a line (in the code block) to prevent the next line from having a prompt. Something similar would be simple to use but I don't know if it's easy to implement. He is using the '\\' character but since its an escape character I guess it may be better to use something else. You know better than me what you can do about that with your plugin.

That's my suggestion.

As for the prompt string changing, just an idea, maybe the use of a special character in the code block could be a way to change the prompt string on the fly. 

**Or** if you have a way to make two code blocks following each other without a gap and keep the line step, we could restart the second code block with a header that specify the new prompt string and the result would be a change of prompt within an apparent single code block.

Like writing this:

````
```bash prompt:"user_name@host:~$" ln:false
cd test \skip_prompt

```

```bash prompt:"user_name@host:~/test$" ln:false connect_to_precedent_block
pwd \skip_prompt
/home/user_name/test
```
````
To get that:

```bash
user_name@host:~$ cd test

user_name@host:~/test$ pwd
/home/user_name/test
```
Of course, see the `\skip_prompt` and `connect_to_precedent_block` as any code word or character you would use instead.

Keep in mind there is no rush for this. These request are goodies not essential functions. 

Thanks

**Comment by mugiwara85** [comment]
@JoseForest If you want a little beta testing, take the `main.js` and the `styles.css` and play with prompts ;) There is a new settingspage in the settings. Simply define `prompt:bash` (predefined values are: bash, bashalt, zsh, zshgit, kali, fish, ps, cmd, docker, postgres). You can also create your own as well e.g. `promp:$` or you could also use templates with user, host, branch, db values. I will document it later in detail before the release. Some basic commands are also supported (cd, pwd, su, whoami, exit, \c). I would appreciate some feedback :) 
(There might be a few smaller bugs, but I think it got pretty good)

**Comment by hyc056** [comment]
@mugiwara85 I briefly tested it and found it quite impressive. It will be very helpful for my learning, and I genuinely hope this feature can be officially released.
```
bash prompt:bash
user@localhost:~/$ cd test
user@localhost:/~/test$ pwd
/~/test
user@localhost:/~/test$ cd var
user@localhost:/~/test/var$ cd ..
user@localhost:/~/test$ cd
user@localhost:~$
```

**Comment by mugiwara85** [comment]
Thanks @hyc056! Yes, it will be released, but I have two other ideas I want to do before the release. I will also document prompts better, because there are a few things you can do with them :)

**Comment by JoseForest** [comment]
@mugiwara85 I did some brief test with bash. It's very good. Beyond what I expected.

I changed the default directory to only "\~". 
Like the code block from hyc056 shows, from the path "\~", when we issue "cd test", a "/" appears at the beginning of the path before the "\~" on the following prompt. But a "cd" alone bring back the correct path "\~" without "/".
Also, from the path "\~", a first "pwd" shows a full path "/home/user" but once we issue a "cd test", "pwd" shows "/~/test" instead of /home/user/test".

I didn't find a way to disable the promt for a line.

Thanks, it will be of great help !

**Comment by mugiwara85** [comment]
@JoseForest Yeah, that is a bug. I just saw that when @hyc056 posted it. I will fix that before release. If the start folder is not ~ then it should work correctly. Btw, following commands should work:
| Command         | Description                                                      |
| --------------- | ---------------------------------------------------------------- |
| `cd`            | Go to home (`~`)                                                 |
| `cd ~`          | Also goes to home                                                |
| `cd folder`     | Go into a folder                                                 |
| `cd ..`         | Go one level up                                                  |
| `cd ../..`      | Go two levels up                                                 |
| `cd -`          | Go back to previous dir                                          |
| `cd "some dir"` | Quoted folder name                                               |
| `cd..`          | Common typo for `cd ..`                                          |
| `cd ~/folder`   | Inside home dir                                                  |
| `cd /absolute`  | Full path                                                        |
| `pwd`           | Shows current working directory                                  |
| `su`            | Switch to root user                                              |
| `su <user>`     | Switch to specific user                                          |
| `whoami`        | Shows the current user                                           |
| `\c`            | Changing database in postgress                                   |
| `exit`          | Switch back to the previous user after using `su` or `su <user>` |
| `git checkout <branch>` | Change branch                                                    |
| `git switch <branch>`   | Change branch                                                    |

You can create prompts "on the fly" e.g. `prompt:$` or with templates e.g. `prompt:[{user} at {host} in {path}] user:test path:/var/www host:machine` You can use user, host, path, branch, db placeholders, and then define them also on the command line.
Note: these on the fly prompt get only the default greyish color. If you want to specify coors for it, then you must create your own in the settings page. Create a new prompt (give it a name e.g. `test`), set the colors, and then you can use it e.g.: `prompt:test`

If you find any bugs, just post them here, and i will fix it before the release.

**Comment by mugiwara85** [comment]
@JoseForest @hyc056 I fixed the path problem. It should be ok now. Could you please take the current `main.js` and `styles.css` and re-test it?

You can also try this. Copy this **exactly** as is, and see the magic happen ;)
````
```group:test
1st code block code
```
```group:test
2nd code block code
```

```group:test2 tab:mytab
1st code block code
```
```group:test2 tab:"another tab"
2nd code block code
```
````

**Comment by mugiwara85** [comment]
Almost forgot. I also added a new option "auto-use prompt". You can define languages for which that specific prompt should be used, and then it will be automatically used. You can disable it with `noprompt` for a specific code block. You can also specify lines or ranges `noprompt:1,3,6-8` for which lines the prompt should **NOT** be used.

**Comment by JoseForest** [comment]
@mugiwara85, the path works fine. The **auto-use prompt** and `noprompt` are very useful. 

The **tabs** are a nice addition. On the PDF export, the tabs are printed one after the other, which allow to see them all.

I tested the commands you listed on May 15th post and everything work fine (at least with a bash prompt) except the `\c` command. 

I tried to change database with PostgreSQL but I could not make it work. 

I tried:
````
```shell prompt:postgres title:"Test4" ln:false
\c Mydb

```
````
And got:

![Image](https://github.com/user-attachments/assets/d7316642-30fd-4093-abcd-addda6b568ea)

I'm not familiar with PostgreSQL. Am I wrong to expect the database to change in the prompt?

Great job! Very helpful.
Thanks

**Comment by mugiwara85** [comment]
@JoseForest Thanks for the feedback! Tabs have one limitation, that when printing, they will always be shown as separate code blocks. I couldn't get it work to render them correctly, so I could add a toggle for it. But actually I was thinking as well, that this isn't actually a big problem.
`\c` should have worked in postgress. The database should have changed. I take a look at it before release. (It did work, but apperantly I broke it somehow :D )

I don't know if you noticed, but you can add syntax highlighting to inline code as well. Just do something like `{python} python code` and it should get the syntax highlighting.

There is one other feature coming up, although I am not yet sure, how I can display it nicely. And there is also some enhancement coming up, which I think is nice to have, but was having trouble implementing it. I think I get it this time.

**Comment by mugiwara85** [comment]
Available in version 1.3.0
